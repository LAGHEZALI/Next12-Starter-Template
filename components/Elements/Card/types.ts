export type CardProps = {
  href?: string;
  title: string;
  content: string;
  variant?: 'primary' | 'secondary';
};
