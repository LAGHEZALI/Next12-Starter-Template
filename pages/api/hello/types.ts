export type Data = {
  message: string;
  data?: string;
};
