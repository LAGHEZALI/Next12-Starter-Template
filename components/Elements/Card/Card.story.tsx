import React from 'react';
import { Story, Meta } from '@storybook/react';

import Card from './index';
import { CardProps } from './types';

export default {
  title: 'Elements/Card',
  component: Card,
  argTypes: {
    variant: {
      control: {
        type: 'select',
        options: ['primary', 'secondary'],
      },
    },
  },
} as Meta;

const Template: Story<CardProps> = (args) => <Card {...args} />;

export const Default = Template.bind({});
Default.args = {
  href: '#',
  title: 'Custom Card',
  content: 'This is the content of a Custom Card.',
};
