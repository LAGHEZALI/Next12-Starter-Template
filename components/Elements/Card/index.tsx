import React from 'react';

import { CardProps } from './types';

import styles from './styles.module.scss';

function Card({ href, title, content, variant = 'primary' }: CardProps) {
  return (
    <div className={`${styles.card} ${variant === 'secondary' ? styles.secondary : styles.primary}`} data-testid="card">
      <a href={href} target="_blank" rel="noopener noreferrer">
        <h2 className={styles.title}>{title}</h2>
        <p className={styles.content}>{content}</p>
      </a>
    </div>
  );
}

export default Card;
