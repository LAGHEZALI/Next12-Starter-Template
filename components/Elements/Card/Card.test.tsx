import React from 'react';
import { render, screen } from '@testing-library/react';

import Card from './index';
import { CardProps } from './types';

describe('Card component', () => {
  const props: CardProps = {
    href: 'https://example.com',
    title: 'Example',
    content: 'This is an example card',
  };

  const secondaryProps: CardProps = {
    ...props,
    variant: 'secondary',
  };

  it('renders the card with the correct props', () => {
    render(<Card {...props} />);
    expect(screen.getByRole('link')).toHaveAttribute('href', props.href);
    expect(screen.getByRole('heading')).toHaveTextContent(props.title);
    expect(screen.getByText(props.content)).toBeInTheDocument();
  });

  it('renders the card with the secndary attributes', () => {
    render(<Card {...secondaryProps} />);
    expect(screen.getByTestId('card')).toHaveClass('secondary');
  });

  it('renders the correct CSS classes', () => {
    render(<Card {...props} />);
    expect(screen.getByRole('heading')).toHaveClass('title');
    expect(screen.getByText(props.content)).toHaveClass('content');
  });

  it('opens the link in a new tab', () => {
    render(<Card {...props} />);
    expect(screen.getByRole('link')).toHaveAttribute('target', '_blank');
    expect(screen.getByRole('link')).toHaveAttribute('rel', 'noopener noreferrer');
  });

  it('matches the snapshot', () => {
    const { asFragment } = render(<Card {...props} />);
    expect(asFragment()).toMatchSnapshot();
  });
});
