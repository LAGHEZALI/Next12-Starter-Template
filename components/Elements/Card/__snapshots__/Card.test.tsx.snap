// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Card component matches the snapshot 1`] = `
<DocumentFragment>
  <div
    class="card primary"
    data-testid="card"
  >
    <a
      href="https://example.com"
      rel="noopener noreferrer"
      target="_blank"
    >
      <h2
        class="title"
      >
        Example
      </h2>
      <p
        class="content"
      >
        This is an example card
      </p>
    </a>
  </div>
</DocumentFragment>
`;
